src/scripts/scaling/scalingAction.sh

```diff
@@ -2,24 +2,272 @@
+  if [ $? -ne 0 ]; then
+    echo "$currdate Failed to retrieve WebLogic Domain Custom Resource Definition" >> scalingAction.log
+    exit 1
+  fi
+  echo "$DOMAIN"
+}
+
+# Verify if cluster is defined in clusters of the Custom Resource Domain
+# args:
+# $1 Custom Resource Domain
+function is_defined_in_clusters() {
+  local DOMAIN="$1"
+  local in_cluster_startup="False"
+  if [ -x "$(command -v jq)" ]; then
+    local inClusterStartupCmd="(.items[].spec.clusters[] | select (.clusterName == \"${wls_cluster_name}\"))"
+    local clusterDefinedInCRD=$(echo "${DOMAIN}" | jq "${inClusterStartupCmd}")
+    if [ "${clusterDefinedInCRD}" != "" ]; then
+      in_cluster_startup="True"
+    fi
+  else
+cat > cmds-$$.py << INPUT
+import sys, json
+outer_loop_must_break = False
+for i in json.load(sys.stdin)["items"]:
+  j = i["spec"]["clusters"]
+  for index, cs in enumerate(j):
+    if j[index]["clusterName"] == "$wls_cluster_name":
+      outer_loop_must_break = True
+      print True
+      break
+if outer_loop_must_break == False:
+  print False
+INPUT
+in_cluster_startup=`echo ${DOMAIN} | python cmds-$$.py`
+  fi
+  echo "$in_cluster_startup"
+}
+
+# Gets the current replica count of the cluster
+function get_num_ms_in_cluster() {
+  local num_ms
+  if [ -x "$(command -v jq)" ]; then
+  local numManagedServersCmd="(.items[].spec.clusters[] | select (.clusterName == \"${wls_cluster_name}\") | .replicas)"
+  num_ms=$(echo "${DOMAIN}" | jq "${numManagedServersCmd}")
+  else
+cat > cmds-$$.py << INPUT
+import sys, json
+for i in json.load(sys.stdin)["items"]:
+  j = i["spec"]["clusters"]
+  for index, cs in enumerate(j):
+    if j[index]["clusterName"] == "$wls_cluster_name":
+      print j[index]["replicas"]
+INPUT
+  num_ms=`echo ${DOMAIN} | python cmds-$$.py`
+  fi
+
+  if [ "${num_ms}" == "null" ]; then
+    num_ms=0
+  fi
+
+  echo "$num_ms"
+}
+
+# Gets the replica count at the Domain level
+function get_num_ms_domain_scope() {
+  local num_ms
+  if [ -x "$(command -v jq)" ]; then
+    num_ms=$(echo "${DOMAIN}" | jq -r '.items[].spec.replicas' )
+  else
+cat > cmds-$$.py << INPUT
+import sys, json
+for i in json.load(sys.stdin)["items"]:
+  print i["spec"]["replicas"]
+INPUT
+  num_ms=`echo ${DOMAIN} | python cmds-$$.py`
+  fi
+
+  if [ "${num_ms}" == "null" ]; then
+    num_ms=0
+  fi
+
+  echo "$num_ms"
+}
+
+# Get the current replica count for the WLS cluster if defined in the CRD's Cluster
+# configuration.  If WLS cluster is not defined in the CRD then return the Domain
+# scoped replica value, if present.  Returns replica count = 0 if no replica count found.
+# args:
+# $1 "True" if WLS cluster configuration defined in CRD, "False" otherwise
+# $2 Custom Resource Domain
+function get_replica_count() {
```

Comment:
Should this take into account 'min replicas' when minimum replicas is enabled?

Reply:
Can you clarify where we should take 'min replicas' into account?  At the moment, we're just scaling up/down based on the currently configured replica count.  I see that the operator will enforce the 'min replicas' at runtime during ManagedServersUpStep.

Reply:
1 - If the 'true' replica count is 2, but this method returns 0, then a scale-up request will scale up to 1, and therefore it won't truly scale-up the cluster.  

2 - I haven't checked if this script already does this already, but I assume it'd be helpful if this script 'failed-fast' on an attempt to scale too low or too high, instead of indirectly relying on an administrator to figure out what went wrong.

The 'life cycle' scripts that Anil recently added have code for getting the min replica count (see getReplicaCount and getMinReplicas 'helper.sh').

Reply:
Added verification of scaling request count against configured minimum replica count for cluster.

Reply:
Also, scaling above configured max cluster size is verified by the REST backend implementation and exception (HTTP status code and message) returned to client.